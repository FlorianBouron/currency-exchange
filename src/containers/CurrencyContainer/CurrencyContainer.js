import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import TextField from "@material-ui/core/TextField";
import { selectors } from "../../redux/wallets";
import {
  setErrorBalanceFrom,
  setErrorBalanceTo,
  selectors as selectorsErrors
} from "../../redux/errors";
import {
  setInputValue,
  selectors as selectorsCurrencies
} from "../../redux/currencies";
import CurrencySelector from "../../components/CurrencySelector";
import styles from "./CurrencyContainer.module.scss";

class CurrencyContainer extends React.PureComponent {
  constructor(props) {
    super(props);
    const { isReadOnly } = props;
    this.signInput = isReadOnly ? "+" : "-";
  }

  setErrorMesages = message => {
    const { isReadOnly } = this.props;
    if (isReadOnly) {
      const { setErrorBalanceTo } = this.props;
      setErrorBalanceTo(message);
    } else {
      const { setErrorBalanceFrom } = this.props;
      setErrorBalanceFrom(message);
    }
  };

  handleChange = ({ target }) => {
    const { amount, isReadOnly, setInputValue } = this.props;
    const { value } = target;
    let inputValue = value;

    if (value === this.signInput) {
      inputValue = "";
    }

    if (Math.abs(value) > amount) {
      const error = "Exceed your balance";
      this.setErrorMesages(error);
    } else {
      this.setErrorMesages("");
    }
    setInputValue(Number(isReadOnly), inputValue);
  };

  handleAllowedCharacters = event => {
    /*
      Table of Values:
      46: .
      48: 0
      57: 9
    */
    if ((event.charCode < 48 || event.charCode > 57) && event.charCode !== 46) {
      event.preventDefault();
    }

    //Prevent to have two dots
    if (event.target.value.indexOf(".") !== -1 && event.charCode === 46) {
      event.preventDefault();
    }

    //Prevent to have more than two digits after the dot
    const stringAfterDot = event.target.value.split(".")[1];
    if (stringAfterDot && stringAfterDot.length >= 2) {
      event.preventDefault();
    }
  };

  render() {
    const {
      wallets,
      className,
      currentCurrency,
      onChangeCurrency,
      error,
      inputValue,
      symbol,
      amount
    } = this.props;
    let displayInputValue = inputValue;
    if (
      this.signInput === "+" &&
      inputValue.toString().indexOf("+") === -1 &&
      inputValue.toString().length
    ) {
      displayInputValue = `${this.signInput}${inputValue}`;
    } else if (this.signInput === "-" && inputValue > 0) {
      displayInputValue = `${this.signInput}${inputValue}`;
    }

    return (
      <div className={[styles["currency-container"], className].join(" ")}>
        <div className={styles["currency-input-selector"]}>
          <CurrencySelector
            currentCurrency={currentCurrency}
            wallets={wallets}
            onChange={onChangeCurrency}
          />
          <TextField
            value={displayInputValue}
            onChange={this.handleChange}
            onKeyPress={this.handleAllowedCharacters}
            type="text"
            placeholder="0"
            className={styles["currency-text-field"]}
            InputLabelProps={{
              shrink: true
            }}
            margin="normal"
          />
        </div>
        <div className={styles["currency-info"]}>
          <div className={error ? [styles["currency-balance-error"]] : null}>
            Balance: {`${symbol}${amount}`}
          </div>
          {error ? <span>{error}</span> : null}
        </div>
      </div>
    );
  }
}

CurrencyContainer.propTypes = {
  wallets: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      amount: PropTypes.number.isRequired
    })
  ),
  currentCurrency: PropTypes.string.isRequired,
  isReadOnly: PropTypes.bool,
  className: PropTypes.string,
  onChangeCurrency: PropTypes.func.isRequired
};

CurrencyContainer.defaultProps = {
  wallets: [],
  isReadOnly: false,
  className: null
};

export default connect(
  (state, props) => ({
    amount: selectors.getWalletByName(state, props.currentCurrency).amount,
    symbol: selectors.getWalletByName(state, props.currentCurrency).symbol,
    error: selectorsErrors.getErrors(state)[
      props.isReadOnly ? "errorBalanceTo" : "errorBalanceFrom"
    ],
    inputValue: selectorsCurrencies.getInputValueByIndex(
      state,
      Number(props.isReadOnly)
    )
  }),
  { setErrorBalanceFrom, setErrorBalanceTo, setInputValue }
)(CurrencyContainer);
