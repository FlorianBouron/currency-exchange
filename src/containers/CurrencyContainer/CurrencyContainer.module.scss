@import "../../constants/utils.scss";
@import "../../constants/colors.scss";

.currency-container {
  padding: px2rem(30px) px2rem(20px);
  .currency-input-selector {
    display: flex;
    align-items: center;
    justify-content: space-between;

    .currency-text-field {
      input {
        text-align: right;
      }
    }
  }
  .currency-info {
    margin: px2rem(12px) 0;
    padding-left: px2rem(25px);
    display: flex;
    justify-content: space-between;
    color: $secondary-gray;
    .currency-balance-error {
      color: $primary-pink;
    }
  }
}
