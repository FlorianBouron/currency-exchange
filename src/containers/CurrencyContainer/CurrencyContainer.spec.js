import React from "React";
import { shallow, mount } from "enzyme";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import NumberFormat from "react-number-format";
import CurrencyContainer from ".";

describe("CurrencyContainer component", () => {
  const middlewares = [thunk];
  const mockStore = configureStore(middlewares);
  const reduxState = {
    currencies: {
      currencyFrom: { name: "GBP", symbol: "£", inputValue: "" },
      currencyTo: { name: "EUR", symbol: "€", inputValue: "" }
    },
    wallets: {
      data: [
        { id: "wallet-id-0", name: "GBP", amount: 200, symbol: "£" },
        { id: "wallet-id-1", name: "EUR", amount: 85, symbol: "€" },
        { id: "wallet-id-2", name: "USD", amount: 1000, symbol: "$" }
      ]
    },
    rates: {
      base: "GBP",
      rates: { USD: 1.2721606179, EUR: 1.1292034599, GBP: 1 },
      error: ""
    },
    errors: { errorBalanceFrom: "" }
  };
  const store = mockStore(reduxState);

  const basicProps = {
    store,
    wallets: [
      { id: "wallet-id-0", name: "GBP", amount: 200, symbol: "£" },
      { id: "wallet-id-1", name: "EUR", amount: 85, symbol: "€" },
      { id: "wallet-id-2", name: "USD", amount: 1000, symbol: "$" }
    ],
    currentCurrency: "EUR",
    className: "App_app__currency-container-read-only__2LQBE",
    onChangeCurrency: jest.fn(),
    currencyRate: "EUR"
  };

  it("has working component", () => {
    const component = shallow(<CurrencyContainer {...basicProps} />);
    expect(component).toBeDefined();
  });

  it("expects component to never be changed", () => {
    const component = shallow(<CurrencyContainer {...basicProps} />);
    expect(component).toMatchSnapshot();
  });

  it("expects component to never be changed with isReadOnly", () => {
    const customReduxState = {
      ...reduxState,
      errors: { errorBalanceFrom: "Some error..." }
    };
    const customStore = mockStore(customReduxState);
    const customProps = {
      ...basicProps,
      store: customStore,
      isReadOnly: true
    };
    const component = shallow(<CurrencyContainer {...customProps} />);
    expect(component).toBeDefined();
    expect(component).toMatchSnapshot();
  });

  it("expects to call setErrorBalanceFrom", () => {
    const customReduxState = {
      ...reduxState,
      currencies: {
        currencyFrom: { name: "GBP", symbol: "£", inputValue: "33" },
        currencyTo: { name: "EUR", symbol: "€", inputValue: "33" }
      },
      wallets: {
        data: [
          { id: "wallet-id-0", name: "GBP", amount: 1, symbol: "£" },
          { id: "wallet-id-1", name: "EUR", amount: 1, symbol: "€" },
          { id: "wallet-id-2", name: "USD", amount: 1000, symbol: "$" }
        ]
      }
    };
    const store = mockStore(customReduxState);

    let setErrorBalanceFromCalled = false;
    const customProps = {
      ...basicProps,
      store,
      setErrorBalanceFrom: () => {
        setErrorBalanceFromCalled = true;
      }
    };
    const component = mount(<CurrencyContainer {...customProps} />);
    expect(component).toBeDefined();
    setTimeout(() => {
      expect(setErrorBalanceFromCalled).toBe(true);
    });
  });

  it("expects to call setErrorBalanceFrom and prevent not allowed character", () => {
    let preventDefaultHasBeenCalled = false;
    const customEvent = {
      charCode: 23,
      preventDefault: () => (preventDefaultHasBeenCalled = true)
    };
    const component = mount(<CurrencyContainer {...basicProps} />);
    component
      .find(NumberFormat)
      .first()
      .simulate("keypress", customEvent);
    expect(preventDefaultHasBeenCalled).toBe(true);
  });

  it("expects to call setErrorBalanceFrom and prevent two dots", () => {
    let preventDefaultHasBeenCalled = false;
    const customEvent = {
      charCode: 46,
      target: {
        value: "200.22"
      },
      preventDefault: () => (preventDefaultHasBeenCalled = true)
    };
    const component = mount(<CurrencyContainer {...basicProps} />);
    component
      .find(NumberFormat)
      .first()
      .simulate("keypress", customEvent);
    expect(preventDefaultHasBeenCalled).toBe(true);
  });

  it("expects to call handleChange and call setErrorBalanceFrom and add error to redux", () => {
    const customEvent = {
      value: 99999
    };
    const component = mount(<CurrencyContainer {...basicProps} />);
    component
      .find(NumberFormat)
      .first()
      .simulate("change", customEvent);
    expect(component).toBeDefined();
  });

  it("expects to call handleChange and call setErrorBalanceFrom and clear error to redux", () => {
    const customEvent = {
      value: 1
    };
    const component = mount(<CurrencyContainer {...basicProps} />);
    component
      .find(NumberFormat)
      .first()
      .simulate("change", customEvent);
    expect(component).toBeDefined();
  });

  it("expects to call handleChange and call setErrorBalanceFrom and clear error to redux", () => {
    const customEvent = {
      value: 1
    };
    const component = mount(<CurrencyContainer {...basicProps} />);
    component
      .find(NumberFormat)
      .first()
      .prop("onValueChange")(customEvent);
    expect(component).toBeDefined();
  });
});
