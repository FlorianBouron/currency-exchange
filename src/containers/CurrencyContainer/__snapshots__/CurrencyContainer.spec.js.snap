// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyContainer component expects component to never be changed 1`] = `
<ContextProvider
  value={null}
>
  <CurrencyContainer
    amount={85}
    className="App_app__currency-container-read-only__2LQBE"
    currencyRate="EUR"
    currentCurrency="EUR"
    error=""
    fetchRates={[Function]}
    inputValue=""
    isReadOnly={false}
    onChangeCurrency={[MockFunction]}
    rate="1.129"
    setErrorBalanceFrom={[Function]}
    setInputValue={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    symbol="€"
    wallets={
      Array [
        Object {
          "amount": 200,
          "id": "wallet-id-0",
          "name": "GBP",
          "symbol": "£",
        },
        Object {
          "amount": 85,
          "id": "wallet-id-1",
          "name": "EUR",
          "symbol": "€",
        },
        Object {
          "amount": 1000,
          "id": "wallet-id-2",
          "name": "USD",
          "symbol": "$",
        },
      ]
    }
  />
</ContextProvider>
`;

exports[`CurrencyContainer component expects component to never be changed with isReadOnly 1`] = `
<ContextProvider
  value={null}
>
  <CurrencyContainer
    amount={85}
    className="App_app__currency-container-read-only__2LQBE"
    currencyRate="EUR"
    currentCurrency="EUR"
    error={null}
    fetchRates={[Function]}
    inputValue=""
    isReadOnly={true}
    onChangeCurrency={[MockFunction]}
    rate="1.129"
    setErrorBalanceFrom={[Function]}
    setInputValue={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    symbol="€"
    wallets={
      Array [
        Object {
          "amount": 200,
          "id": "wallet-id-0",
          "name": "GBP",
          "symbol": "£",
        },
        Object {
          "amount": 85,
          "id": "wallet-id-1",
          "name": "EUR",
          "symbol": "€",
        },
        Object {
          "amount": 1000,
          "id": "wallet-id-2",
          "name": "USD",
          "symbol": "$",
        },
      ]
    }
  />
</ContextProvider>
`;
